File: apps/web/src/types/attendance.ts
// Minimal interfaces for relations
export interface Student {
    id: number;
    name: string;
    admission_number: string;
}

export interface Faculty {
    id: number;
    user: { full_name: string };
}

export interface Subject {
    id: number;
    name: string;
    code: string;
}

export enum AttendanceStatus {
    PRESENT = "PRESENT",
    ABSENT = "ABSENT",
    LATE = "LATE",
    ON_DUTY = "ON_DUTY"
}

export enum SessionStatus {
    SCHEDULED = "SCHEDULED",
    COMPLETED = "COMPLETED",
    CANCELLED = "CANCELLED"
}

export interface AttendanceSession {
    id: number;
    subject_id: number;
    faculty_id: number;
    program_id: number;
    program_year_id: number;
    semester: number;
    section: string;
    session_date: string; // ISO date string
    start_time: string; // ISO time string
    end_time: string; // ISO time string
    topic_covered?: string;
    status: SessionStatus;
    created_at: string;
    updated_at: string;
    subject?: Subject;
    faculty?: Faculty;
}

export interface AttendanceRecord {
    id: number;
    session_id: number;
    student_id: number;
    status: AttendanceStatus;
    remarks?: string;
    student?: Student;
}

export interface AttendanceStats {
    total_classes: number;
    present: number;
    absent: number;
    late: number;
    on_duty: number;
    attendance_percentage: number;
}

export interface CreateSessionDTO {
    subject_id: number;
    faculty_id: number;
    program_id: number;
    program_year_id: number;
    semester: number;
    section: string;
    session_date: string;
    start_time: string;
    end_time: string;
    topic_covered?: string;
}

export interface UpdateRecordDTO {
    student_id: number;
    status: AttendanceStatus;
    remarks?: string;
}

export interface BulkMarkAttendanceDTO {
    session_id: number;
    records: UpdateRecordDTO[];
}

File: apps/api/app/models/attendance.py
from datetime import date, time, datetime
from typing import Optional, List
from enum import Enum
import sqlalchemy as sa
from sqlmodel import Field, Relationship, SQLModel

# Enums
class AttendanceStatus(str, Enum):
    PRESENT = "PRESENT"
    ABSENT = "ABSENT"
    LATE = "LATE"
    ON_DUTY = "ON_DUTY"
    
class SessionStatus(str, Enum):
    SCHEDULED = "SCHEDULED"
    COMPLETED = "COMPLETED"
    CANCELLED = "CANCELLED"

# Models
class AttendanceSession(SQLModel, table=True):
    __tablename__ = "attendance_session"
    
    id: Optional[int] = Field(default=None, primary_key=True)
    subject_id: int = Field(foreign_key="subject.id")
    faculty_id: int = Field(foreign_key="faculty.id")
    program_id: int = Field(foreign_key="program.id")
    program_year_id: int = Field(foreign_key="program_years.id")
    
    semester: int
    semester: int
    section: Optional[str] = Field(default=None, max_length=10) # Made optional
    practical_batch_id: Optional[int] = Field(default=None, foreign_key="practical_batch.id")
    session_date: date
    start_time: time
    end_time: time
    
    status: SessionStatus = Field(default=SessionStatus.SCHEDULED)
    topic_covered: Optional[str] = Field(default=None, sa_column=sa.Column(sa.Text))
    
    created_at: datetime = Field(default_factory=datetime.now)
    updated_at: datetime = Field(default_factory=datetime.now, sa_column_kwargs={"onupdate": datetime.now})

    # Relationships
    subject: Optional["Subject"] = Relationship()
    faculty: Optional["Faculty"] = Relationship()
    # program: Optional["Program"] = Relationship()  # Add if Program model is available
    attendance_records: List["AttendanceRecord"] = Relationship(back_populates="session")


class AttendanceRecord(SQLModel, table=True):
    __tablename__ = "attendance_record"
    
    id: Optional[int] = Field(default=None, primary_key=True)
    session_id: int = Field(foreign_key="attendance_session.id", ondelete="CASCADE")
    student_id: int = Field(foreign_key="student.id")
    
    status: AttendanceStatus
    remarks: Optional[str] = None
    
    created_at: datetime = Field(default_factory=datetime.now)
    updated_at: datetime = Field(default_factory=datetime.now, sa_column_kwargs={"onupdate": datetime.now})
    
    # Relationships
    session: AttendanceSession = Relationship(back_populates="attendance_records")
    student: Optional["Student"] = Relationship()
